### PR TITLE
Implement caching of vtables, byte-slices and strings

### LINKED ISSUE #20: [Rust] Implement optional caching of vtables
Ideally we want pluggable backends where you can pick if you want caching or not.

### PR DESCRIPTION
**Pull request by TethysSvensson**
**Checklist**
- [x] Updated CHANGELOG.md with relevant changes
- [x] Added tests for any new/fixed functionality
- [x] Added/updated documentation for new/changed code
- [x] Checked that README.md still makes sense (and updated it if necessary)

This fixes #20.

### CODE REVIEW

**Review by kristoff3r — APPROVED** [review]

File: planus/src/builder_cache.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+fn hash_one<H: BuildHasher>(hash_builder: &H, value: &[u8]) -> u64 {
+    let mut hasher = hash_builder.build_hasher();
+    value.hash(&mut hasher);
+    hasher.finish()
+}
+
+impl<C: Cacheable> Cache<C> {
+    pub(crate) fn hash(&self, serialized_data: &[u8]) -> u64 {
```
[the remark this answers is not in the document]
Comment (TethysSvensson) [reply planus/src/builder_cache.rs:60]: Used in more places than I thought, inlining does not make sense.

File: planus/src/table_writer.rs [hunk 2]
```diff
@@ -20,33 +19,45 @@ impl<'buf, const VTABLE_MAX_BYTES: usize, const OBJECT_MAX_BYTES: usize>
...
     pub fn finish_calculating(&mut self) {
-        self.buffer_position = self.builder.get_buffer_position_and_prepare_write(
-            self.vtable_size,
-            self.object_size,
-            self.object_alignment_mask,
-        );
+        self.vtable[0..2].copy_from_slice(&(self.vtable_size as u16).to_le_bytes());
+        self.vtable[2..4].copy_from_slice(&(self.object_size as u16).to_le_bytes());
+
+        let vtable_offset = self.builder.write_vtable(&self.vtable[..self.vtable_size]);
```
[the remark this answers is not in the document]
Comment (TethysSvensson) [reply planus/src/table_writer.rs]: I did something better

File: planus/src/builder_cache.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+use hashbrown::raw::RawTable;
+
+pub(crate) trait Cacheable {
+    fn lookup(serialized: &[u8]) -> Option<&[u8]>;
+}
+
+pub(crate) struct VTable;
+
+impl Cacheable for VTable {
+    fn lookup(serialized: &[u8]) -> Option<&[u8]> {
```
[the remark this answers is not in the document]
Comment (TethysSvensson) [reply planus/src/builder_cache.rs]: Done

File: planus/src/builder_cache.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+impl Cacheable for ByteVec {
+    fn lookup(serialized: &[u8]) -> Option<&[u8]> {
+        let length = u32::from_le_bytes(serialized.get(..4)?.try_into().ok()?);
+        serialized.get(4..4 + length as usize)
+    }
+}
+
+pub(crate) struct Cache<C> {
+    _marker: PhantomData<C>,
+    cache: RawTable<u32>,
```
[the remark this answers is not in the document]
Comment (TethysSvensson) [reply planus/src/builder_cache.rs]: Done

[the remark this answers is not in the document]
Comment (TethysSvensson) [reply planus/src/builder_cache.rs]: Done

File: planus/src/builder.rs [hunk 5]
```diff
@@ -205,20 +244,29 @@ impl Builder {
...
         if let Some(file_identifier) = file_identifier {
             // TODO: how does alignment interact with file identifiers? Is the alignment with out without the header?
-            self.prepare_write(
+            let offset = self.prepare_write(
                 8,
                 <Offset<T> as Primitive>::ALIGNMENT_MASK.max(self.alignment_mask),
-            );
-            self.write(&(4 + self.inner.len() as u32 - root.offset).to_le_bytes());
+            ) as u32;
+            self.write(&(offset - 4 - root.offset).to_le_bytes());
```
[the remark this answers is not in the document]
Comment (TethysSvensson) [reply planus/src/builder.rs:250]: I have added #141 for this.

**Pull request merged**